runtime/ast/entitlement_declaration.go

```diff
@@ -0,0 +1,282 @@
+}
+
+func (d *EntitlementDeclaration) DeclarationDocString() string {
+	return d.DocString
+}
+
+func (d *EntitlementDeclaration) MarshalJSON() ([]byte, error) {
+	type Alias EntitlementDeclaration
+	return json.Marshal(&struct {
+		*Alias
+		Type string
+	}{
+		Type:  "EntitlementDeclaration",
+		Alias: (*Alias)(d),
+	})
+}
+
+var entitlementKeywordSpaceDoc = prettier.Text("entitlement ")
+
+func (d *EntitlementDeclaration) Doc() prettier.Doc {
+	var doc prettier.Concat
+
+	if d.Access != AccessNotSpecified {
+		doc = append(
+			doc,
+			prettier.Text(d.Access.Keyword()),
+			prettier.Space,
+		)
+	}
+
+	doc = append(
+		doc,
+		entitlementKeywordSpaceDoc,
+		prettier.Text(d.Identifier.Identifier),
+	)
+
+	return doc
+}
+
+func (d *EntitlementDeclaration) String() string {
+	return Prettier(d)
+}
+
+type EntitlementMapElement struct {
+	Input  *NominalType
+	Output *NominalType
+}
+
+func NewEntitlementMapElement(
+	gauge common.MemoryGauge,
+	input *NominalType,
+	output *NominalType,
+) *EntitlementMapElement {
+	common.UseMemory(gauge, common.EntitlementMappingElementMemoryUsage)
+
+	return &EntitlementMapElement{
+		Input:  input,
+		Output: output,
+	}
+}
+
+var arrowKeywordSpaceDoc = prettier.Text(" -> ")
+
+func (d EntitlementMapElement) Doc() prettier.Doc {
+	var doc prettier.Concat
+
+	return append(
+		doc,
+		d.Input.Doc(),
+		arrowKeywordSpaceDoc,
+		d.Output.Doc(),
+	)
+}
+
+// EntitlementMappingDeclaration
+type EntitlementMappingDeclaration struct {
+	Access       Access
+	DocString    string
+	Identifier   Identifier
+	Associations []*EntitlementMapElement
+	Range
+}
+
+var _ Element = &EntitlementMappingDeclaration{}
+var _ Declaration = &EntitlementMappingDeclaration{}
+var _ Statement = &EntitlementMappingDeclaration{}
+
+func NewEntitlementMappingDeclaration(
+	gauge common.MemoryGauge,
+	access Access,
+	identifier Identifier,
+	associations []*EntitlementMapElement,
+	docString string,
+	declRange Range,
+) *EntitlementMappingDeclaration {
+	common.UseMemory(gauge, common.EntitlementDeclarationMemoryUsage)
+
+	return &EntitlementMappingDeclaration{
+		Access:       access,
+		Identifier:   identifier,
+		Associations: associations,
+		DocString:    docString,
+		Range:        declRange,
+	}
+}
+
+func (*EntitlementMappingDeclaration) ElementType() ElementType {
+	return ElementTypeEntitlementDeclaration
+}
+
+func (*EntitlementMappingDeclaration) Walk(_ func(Element)) {}
+
+func (*EntitlementMappingDeclaration) isDeclaration() {}
+
+// NOTE: statement, so it can be represented in the AST,
+// but will be rejected in semantic analysis
+func (*EntitlementMappingDeclaration) isStatement() {}
+
+func (d *EntitlementMappingDeclaration) DeclarationIdentifier() *Identifier {
+	return &d.Identifier
+}
+
+func (d *EntitlementMappingDeclaration) DeclarationAccess() Access {
+	return d.Access
+}
+
+func (d *EntitlementMappingDeclaration) DeclarationKind() common.DeclarationKind {
+	return common.DeclarationKindEntitlement
```

Comment:
Probably should have its own declaration kind, `DeclarationKindEntitlementMapping`